sql/database.go

```diff
@@ -294,16 +369,187 @@ func Open(uri string, opts ...Opt) (*sqliteDatabase, error) {
+func moveMigratedDB(config *conf, fromPath, toPath string) (err error) {
+	config.logger.Warn("finalizing migration by moving the temporary DB to the original path",
+		zap.String("fromPath", fromPath),
+		zap.String("toPath", toPath))
+	// Try to open the temporary migrated DB in exclusive mode before deleting the
+	// original one.
+	// If the temporary DB is being copied to the original path by another
+	// process, this will fail and the original database will not be deleted.
+	// We don't use the proper database schema here because the temporary DB
+	// may have been created with a different set of migrations.
+	db, err := Open("file:"+fromPath,
+		WithLogger(config.logger),
+		WithConnections(1),
+		WithTemp(),
+		WithNoCheckSchemaDrift(),
+		WithExclusive(),
+	)
+	if err != nil {
+		return fmt.Errorf("open temporary DB %s: %w", fromPath, err)
+	}
+	if err := deleteDB(toPath); err != nil {
+		return err
+	}
+	if err := db.vacuumInto(toPath); err != nil {
+		db.Close()
+		return err
+	}
+	// Open the freshly vacuumed DB in exclusive mode to avoid race condition when
+	// another process also tries to vacuum the temporary DB into the original path
+	// after we close the temporary DB.
+	origDB, err := Open("file:"+toPath,
+		WithLogger(config.logger),
+		WithConnections(1),
+		WithMigrationsDisabled(),
+		WithNoCheckSchemaDrift(),
+		withDisableIncompleteMigrationHandling(),
+		WithExclusive(),
+	)
+	if err != nil {
+		return fmt.Errorf("open vacuumed DB %s: %w", toPath, err)
+	}
+	if err := db.Close(); err != nil {
+		origDB.Close()
+		return fmt.Errorf("close temporary DB %s: %w", fromPath, err)
+	}
+	if err := deleteDB(fromPath); err != nil {
+		origDB.Close()
+		return err
+	}
+	if err := origDB.Close(); err != nil {
+		return fmt.Errorf("close DB %s after migration: %w", toPath, err)
+	}
+	return nil
 }
 
+func dbMigrationPaths(uri string) (dbPath, migratedPath string, err error) {
+	url, err := url.Parse(uri)
+	if err != nil {
+		return "", "", fmt.Errorf("parse uri: %w", err)
+	}
+	if url.Scheme != "file" {
+		return "", "", nil
+	}
+	path := url.Opaque
+	if path == "" {
+		path = url.Path
+	}
+	return path, path + "_migrate", nil
+}
+
+// handleIncompleteCopyMigration handles incomplete copy-based migrations.
+// It only works for 'file:' URIs, doing nothing for other URIs.
+// It first checks if there's a copy of the database with "_migrate" suffix.
+// If it's there, it checks if the migration is complete by checking if
+// DBNAME_migrate_done file exists. It it doesn't, the migration is considered
+// incomplete and the migrated database is removed. If DBNAME_migrate_done
+// file exists, the migration is finalized by running "VACUUM INTO" on the
+// migrated database and replacing the original, after which the migrated
+// database is deleted.
+func handleIncompleteCopyMigration(config *conf) error {
+	dbPath, migratedPath, err := dbMigrationPaths(config.uri)
+	if err != nil {
+		return err
```

Comment:
would be nice to have error instrumentation here

Reply:
Fixed